alf/experience_replayers/segment_tree.py

```diff
@@ -0,0 +1,227 @@
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""SegmentTree."""
+
+import math
+import torch
+import torch.nn as nn
+
+import alf
+from alf.nest.utils import convert_device
+
+
+class SegmentTree(nn.Module):
+    """
+    Data structure to allow efficient calculate the summary statistics over a
+    segment of elements.
+    See https://en.wikipedia.org/wiki/Segment_tree for detail.
+
+    In this implementation, ``values[1]`` is the root. ``values[capacity: 2*capacity]``
+    are the leaves. The two children of an internal node ``values[i]`` are ``values[2*i]``
+    and ``values[2*i+1]``. And ``values[i]`` is set to ``op(values[2*i], values[2*i+1])``.
+    Each leaf represent a value set through ``__setitem__``. All the nodes of
+    tree are initialized to be zeros.
+    """
+
+    def __init__(self,
+                 capacity,
+                 op,
+                 dtype=torch.float32,
+                 device="cpu",
+                 name="SegmentTree"):
+        super().__init__()
+        self._name = name
+        self._device = device
+        with alf.device(self._device):
+            self.register_buffer("_values",
+                                 torch.zeros((2 * capacity, ), dtype=dtype))
+        self._op = op
+        self._capacity = capacity
+        self._leftmost_leaf = 1
+        self._depth = 0
+        while self._leftmost_leaf < capacity:
+            self._leftmost_leaf *= 2
+            if self._leftmost_leaf < capacity:
+                self._depth += 1
+
+    def __setitem__(self, indices, values):
+        """Set the value of leaves and update the internal nodes.
+
+        Args:
+            indices (Tensor): 1-D int64 Tensor. Its values should be in range
+                [0, capacity).
+            values (Tensor): 1-D Tensor with the same shape as ``indices``
+        """
+
+        def _step(indices):
+            """
+            Calculate the parent value from its children.
+            """
+            indices = indices / 2
+            indices = torch.unique(indices)
+            left = self._values[indices * 2]
+            right = self._values[indices * 2 + 1]
+            self._values[indices] = op(left, right)
+            return indices
+
+        with alf.device(self._device):
+            indices = convert_device(indices)
+            values = convert_device(values)
+
+            assert indices.ndim == 1
+            assert values.ndim == 1
+            assert indices.shape == values.shape, (
+                "indices and values should be 1-D tensor with the length. Got "
+                "%s and %s." % (indices.shape, values.shape))
+            op = self._op
```

Comment:
remove this line and use self._op in _step?

Reply:
Python does not do compile optimization. So sometime I do optimization for python.

Reply:
Haha interesting.